packages/widgets/src/utils.js

```diff
@@ -31,3 +35,32 @@ export function addWidgetIdToBlock( block, widgetId ) {
 		},
 	};
 }
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `__internalWidgetId` if needed.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Updated block settings.
+ */
+function addInternalWidgetIdAttribute( settings ) {
+	return {
+		...settings,
+		attributes: {
+			...settings.attributes,
```

Comment:
What’s the reason it gets injected with a filter rather than included in the block settings? It would be great to document it in this callback so it doesn’t get accidentally inlined in `block.json`.

Reply:
The filter lets us inject the `widgetId` in the appropriate context, rather than to all blocks even outside of the widget editor. This used to be absolutely necessary because we used the existence of a `widgetId` attribute to determine whether a block was a widget, but I actually think that's not strictly necessary after[ these changes](https://github.com/WordPress/gutenberg/pull/34750/commits/39eaea7ffd71a50cd8009e6ffd087639202fa1ed).

Reply:
> but I actually think that's not strictly necessary after these changes.

That link doesn't work for me anymore. I imagine it's still necessary that we add `widgetId` to all blocks using this filter because `widgetId` is how the widgets editor knows which widget entity to update when a block is modified.

Reply:
Sorry, link broken after a force push :) [Updated link,](https://github.com/WordPress/gutenberg/pull/34750/commits/0c4673b7a1e74d9414aec63146cd8c9c10afb068) although I think that point is moot -- I meant that it used to be necessary that `widgetId` only be applied to blocks that actually map to a widget entity rather than to all blocks, but really it only matters if the attribute has a value set.

Regardless, I think it's better to add `widgetId` using this filter than automatically to all blocks regardless of context, unless I misunderstand.